### PR TITLE
port of wordy exercism

### PR DESCRIPTION
**Pull request by yanick**

### CODE REVIEW

**Comment by kytrinyx** [comment]
One of the three versions of the build is failing with:

```
# Looks like you planned 16 tests, but ran 5
Dubious, test returned 255 (wstat 65280, 0xff00)
Failed 11/16 subtests 
Test Summary Report
-------------------
exercises/wordy/wordy.t                        (Wstat: 65280 Tests: 5 Failed: 0)
  Non-zero exit status: 255
  Parse errors: Bad plan.  You planned 16 tests but ran 5.
Files=13, Tests=141, 11 wallclock secs ( 0.07 usr  0.02 sys +  9.78 cusr  1.21 csys = 11.08 CPU)
Result: FAIL
The command "EXERCISM=1 prove -vr --exec=perl6 exercises/" exited with 1.
```

**Comment by mienaikage** [comment]
We seem to be encountering problems with 2016.07 quite a bit. Sometimes we even get builds that fail for no apparent reason, so I'm considering removing it and encouraging users to stick with the latest version.

File: exercises/wordy/wordy.t [hunk 1]
```diff
@@ -0,0 +1,26 @@
+#!/usr/bin/env perl6
+
+use Test;
+use JSON::Tiny;
+
+use lib IO::Path.new($?FILE).parent.path;
+
+
+my $module = %*ENV<EXERCISM>.so ?? 'Example.pm' !! 'Wordy.pm';
```
Comment (mienaikage) [inline exercises/wordy/wordy.t]: Could this be changed to allow other extensions e.g. pm6?

File: exercises/wordy/wordy.t [hunk 2]
```diff
@@ -0,0 +1,26 @@
+#!/usr/bin/env perl6
+
+use Test;
+use JSON::Tiny;
+
+use lib IO::Path.new($?FILE).parent.path;
```
Comment (mienaikage) [inline exercises/wordy/wordy.t]: No need to change it but just FYI you can do `(my $dir IO::Path.new($?FILE).parent).path` and then reuse the IO::Path object for the directory later :)

Comment (yanick) [inline exercises/wordy/wordy.t]: True, true.

**Comment by yanick** [comment]
[failing on the rakuda of July] Yeah, considering that Perl6 is so new and its interpreter still under lot of churn, I tend to agree with mienaikage. If it's working with the latest stable rakudo, it ought to be gud enuff(tm).

File: exercises/wordy/wordy.t (hunk 1 above)
Comment (yanick) [inline exercises/wordy/wordy.t]: Dün.

**Comment by kotp** [comment]
Should we use 6c and leave it at that?  At least for now.  It hasn't even been a year yet, right?

**Comment by mienaikage** [comment]
The versions in the .travis.yml are releases of the Perl 6 compiler Rakudo. They are all implementations of Perl v6.c. Keeping up with the latest version of the compiler is probably the best way to go while bug fixing and feature implementation continues.

**Comment by kotp** [comment]
The contents of config.json seem to be the same as the canonical data in x-common.

Shouldn't it be read in and used?

**Comment by yanick** [comment]
> Shouldn't it be read in and used?

I'm not sure I understand. It _is_ read and used by the test file.

**Comment by kotp** [comment]
Read in from exerciscism/x-common rather than hard coded here.  Sorry, I was unclear.

**Comment by yanick** [comment]
Oh, All the exercisms I saw so far are copying the cases in the local repo. Is there an expectation that the x-common repo is  going to be checkout out alongside the exercism repo? Or should it be a submodule?

**Comment by kotp** [comment]
Well, I don't submodule it for the Ruby track, I just have it cloned next to the xruby track.  (There is some documentation about this).  Then we read the config that way.

**Comment by yanick** [comment]
> There is some documentation about this

A link pointing to that documentation would be fantastic.

**Comment by kotp** [comment]
Indeed.  I have to find it, it has been a long time since I have seen it.

**Comment by kotp** [comment]
Found it, it is the [Readme](https://github.com/exercism/xruby/blob/master/README.md) file in the xruby track.

**Comment by yanick** [comment]
Okay, changed the code so it expects `x-common` to be checked out beside the repo. I'm not too sure that's really what we want, but hey, as requested, so it is implemented. :-)

File: exercises/wordy/wordy.t [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
+my $module = first { $dir.child($_).e }, |@potential_module
+    or die "No file '$module_name.p6' found\n";
+
+require $module <&answer>;
+
+plan 16;
+
+# we expect x-common to be checked out alongside xperl6
+my %cases = from-json $dir.parent.parent.parent.child('x-common/exercises/wordy/canonical-data.json').slurp;
```
Comment (kotp) [inline exercises/wordy/wordy.t]: This seems to be failing.  Do we need to add in relative folder reference?

**Comment by kotp** [comment]
In doing so, we should now be able to use the repository locally to test, by reading in the canonical data.  On the server, it will have that asset available, so it will work the same way.

Currently failing checks though.  Perhaps relative folder reference missing? `../x-common/exercises/wordy/....`

**Comment by mienaikage** [comment]
We cannot have the test suite look outside the exercise directory for test data, that isn't going to work when the user fetches the exercise.

I'm not experienced with Ruby, but it looks like the xruby track has template toolkit files (named 'example' so they're not fetched by the user) which is used to generate a self contained test suite. We don't have anything like that for this track yet, though.

**Comment by kotp** [comment]
Yes, it is true that everything is self contained when the user fetches the exercise.  The test file is generated from the data, the developer that pushes the solution to the track runs a tool that is written to read that data from outside of the repository.

The test cases would then be written inside the test suite themselves.

What this means for the track though, is that if the canonical data changes, it is simply running the generator, confirming the work, and pushing up.

Of course, you are right, the reference can't be in that file, but should be in a generator file.  And from the exercise folder everything contained.

**Comment by yanick** [comment]
To remove the extra-repo testcase fetching, you can revert to 0b055fbc0a577804da39a510b2444a77b9224cbb

**Comment by kotp** [comment]
You can checkout to 0b055fb and force push to your branch @yanick and it will do the expected.

**Comment by yanick** [comment]
done

**Pull request merged**